### ISSUE TITLE
sync_file_range02 failed

### ISSUE THREAD

**Issue by josephhz** [linux-test-project/ltp#572]
Hi, I've got the following failed message when running case sync_file_range02:
sync_file_range02.c:74: FAIL: Sync inside of write: Synced 22282240, expected 16777216
sync_file_range02.c:74: FAIL: Sync overlaps with write: Synced 9699328, expected 8388608

Before commit b928910ca769 "syscalls/sync_file_range: add partial file sync test-cases", the case is PASS. I've noticed that this commit has changed the PASS condition:
Before: written >= FILE_SIZE
After: (written >= tc->exp_sync_size) && (written <= (tc->exp_sync_size + tc->exp_sync_size/10))

I'm not sure why we set a threshold 10%. And I think it has missed some scenarios.

My kernel is 4.19.67, using ext4, and this issue can only be found in debug kernel.

**Comment by metan-ucw** [comment]
Can you include some details here? Such as distribution/kernel version the test output when it failed etc. We can hardly do anything without this information.

**Comment by casparant** [comment]
> Can you include some details here? Such as distribution/kernel version the test output when it failed etc. We can hardly do anything without this information.

Hi Cyril,

We use Alibaba Cloud Linux distro, a CentOS 7-like distribution with [4.19 LTS-based kernel](https://alibaba.github.io/cloud-kernel/) (more precisely, v4.19.67-based for now).

the test was based on latest LTP tree, and the full output is:

```
<<<test_output>>>
tst_device.c:238: INFO: Using test device LTP_DEV='/dev/loop0'
tst_supported_fs_types.c:60: INFO: Kernel supports ext2
tst_supported_fs_types.c:44: INFO: mkfs.ext2 does exist
tst_supported_fs_types.c:60: INFO: Kernel supports ext3
tst_supported_fs_types.c:44: INFO: mkfs.ext3 does exist
tst_supported_fs_types.c:60: INFO: Kernel supports ext4
tst_supported_fs_types.c:44: INFO: mkfs.ext4 does exist
tst_supported_fs_types.c:60: INFO: Kernel supports xfs
tst_supported_fs_types.c:44: INFO: mkfs.xfs does exist
tst_supported_fs_types.c:60: INFO: Kernel supports btrfs
tst_supported_fs_types.c:44: INFO: mkfs.btrfs does exist
tst_supported_fs_types.c:60: INFO: Kernel supports vfat
tst_supported_fs_types.c:44: INFO: mkfs.vfat does exist
tst_supported_fs_types.c:83: INFO: Filesystem exfat is not supported
tst_supported_fs_types.c:83: INFO: Filesystem ntfs is not supported
tst_test.c:1161: INFO: Testing on ext2
tst_mkfs.c:90: INFO: Formatting /dev/loop0 with ext2 opts='' extra opts=''
mke2fs 1.43.5 (04-Aug-2017)
tst_test.c:1100: INFO: Timeout per run is 0h 05m 00s
sync_file_range02.c:71: PASS: Sync equals write
sync_file_range02.c:71: PASS: Sync inside of write
sync_file_range02.c:71: PASS: Sync overlaps with write
tst_test.c:1161: INFO: Testing on ext3
tst_mkfs.c:90: INFO: Formatting /dev/loop0 with ext3 opts='' extra opts=''
mke2fs 1.43.5 (04-Aug-2017)
tst_test.c:1100: INFO: Timeout per run is 0h 05m 00s
sync_file_range02.c:71: PASS: Sync equals write
sync_file_range02.c:74: FAIL: Sync inside of write: Synced 25457664, expected 16777216
sync_file_range02.c:71: PASS: Sync overlaps with write
tst_test.c:1161: INFO: Testing on ext4
tst_mkfs.c:90: INFO: Formatting /dev/loop0 with ext4 opts='' extra opts=''
mke2fs 1.43.5 (04-Aug-2017)
tst_test.c:1100: INFO: Timeout per run is 0h 05m 00s
sync_file_range02.c:74: FAIL: Sync equals write: Synced 40115200, expected 33554432
sync_file_range02.c:71: PASS: Sync inside of write
sync_file_range02.c:71: PASS: Sync overlaps with write
tst_test.c:1161: INFO: Testing on xfs
tst_mkfs.c:90: INFO: Formatting /dev/loop0 with xfs opts='' extra opts=''
tst_test.c:1100: INFO: Timeout per run is 0h 05m 00s
sync_file_range02.c:71: PASS: Sync equals write
sync_file_range02.c:71: PASS: Sync inside of write
sync_file_range02.c:71: PASS: Sync overlaps with write
tst_test.c:1161: INFO: Testing on btrfs
tst_mkfs.c:90: INFO: Formatting /dev/loop0 with btrfs opts='' extra opts=''
tst_test.c:1100: INFO: Timeout per run is 0h 05m 00s
sync_file_range02.c:71: PASS: Sync equals write
sync_file_range02.c:71: PASS: Sync inside of write
sync_file_range02.c:71: PASS: Sync overlaps with write
tst_test.c:1161: INFO: Testing on vfat
tst_mkfs.c:90: INFO: Formatting /dev/loop0 with vfat opts='' extra opts=''
tst_test.c:1100: INFO: Timeout per run is 0h 05m 00s
sync_file_range02.c:111: INFO: Pre-filling file
sync_file_range02.c:71: PASS: Sync equals write
sync_file_range02.c:71: PASS: Sync inside of write
sync_file_range02.c:71: PASS: Sync overlaps with write

Summary:
passed   16
failed   2
skipped  0
warnings 0
<<<execution_status>>>
initiation_status="ok"
duration=239 termination_type=exited termination_id=1 corefile=no
cutime=7 cstime=15654
<<<test_end>>>
```

I will assign this issue to myself and take a look at it.

**Comment by casparant** [comment]
looks like something related to 

```
commit c553ea4fdf2701d64b9e9cca4497a8a2512bb025
Author: Amir Goldstein <amir73il@gmail.com>
Date:   Mon May 13 17:22:30 2019 -0700

    fs/sync.c: sync_file_range(2) may use WB_SYNC_ALL writeback

    23d0127096cb ("fs/sync.c: make sync_file_range(2) use WB_SYNC_NONE
    writeback") claims that sync_file_range(2) syscall was "created for
    userspace to be able to issue background writeout and so waiting for
    in-flight IO is undesirable there" and changes the writeback (back) to
    WB_SYNC_NONE.

    This claim is only partially true.  It is true for users that use the flag
    SYNC_FILE_RANGE_WRITE by itself, as does PostgreSQL, the user that was the
    reason for changing to WB_SYNC_NONE writeback.

    However, that claim is not true for users that use that flag combination
    SYNC_FILE_RANGE_{WAIT_BEFORE|WRITE|_WAIT_AFTER}.  Those users explicitly
    requested to wait for in-flight IO as well as to writeback of dirty pages.

    Re-brand that flag combination as SYNC_FILE_RANGE_WRITE_AND_WAIT and use
    WB_SYNC_ALL writeback to perform the full range sync request.

    Link: http://lkml.kernel.org/r/20190409114922.30095-1-amir73il@gmail.com
    Link: http://lkml.kernel.org/r/20190419072938.31320-1-amir73il@gmail.com
    Fixes: 23d0127096cb ("fs/sync.c: make sync_file_range(2) use WB_SYNC_NONE")
    Signed-off-by: Amir Goldstein <amir73il@gmail.com>
    Acked-by: Jan Kara <jack@suse.com>
    Cc: Dave Chinner <david@fromorbit.com>
    Cc: Al Viro <viro@zeniv.linux.org.uk>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
```

This commit missed 4.19 LTS tree. I'll dig a little more....

**Comment by xuyang0410** [comment]
10% limit has been remove since [sync_file_range02: remove the toplimit of write back ](https://github.com/linux-test-project/ltp/commit/1643d850911c7d6dbf898e1e3dfeef974e84508b) because it may existed a delay io before sync on ext4 and we call sync [tst_device: add new tst_dev_sync ](https://github.com/linux-test-project/ltp/commit/eca0fa3c3aa116808cc60d206625c9f8a7bbfa36) before count to avoid this situation. So now we can consider it success when greater than expected btyes.
